# Describe how to run program

## Perl dependencies
 List::MoreUtils (https://metacpan.org/pod/List::MoreUtils)

## Sample run
The associated dependencies of APP must be installed prior to attempt running the program. The details of these have been provided in [Installation.md](https://github.com/sohamsg90/APP-Alieness-by-Phyletic-Pattern/blob/main/docs/Installation.md). 

## Input
We are using a random genome *Yersinia pestis KIM10+* to perform our sample run . The NCBI accession number is  [NC_004088] (https://www.ncbi.nlm.nih.gov/nuccore/NC_004088.1/).

APP requires two input files - 
1. A file with taxonomic IDs of the associated genome
2. A multiFASTA protein sequence file

### Constructing taxonomic ID file

The taxonomic IDs can be generated by using the [`input_file_prepare.pl`] (https://github.com/sohamsg90/APP-Alienness-by-Phyletic-Pattern/blob/main/scripts/taxonomy/input_file_prepare.pl) script. An input file with the genome accession number (NCBI) is to be provided like,

`perl scripts/taxonomy/input_file_prepare.pl accession_number.txt`

This generates a file `3.query_speciesID_taxID.txt` with the species ID and taxonomic ID of the genome (from NCBI taxonomy database).

The output of the file looks like this:
| Acc       | speciesid | taxid  |
| --------- | --------- | ------ |
| NC_004088 | 632       | 187410 |

This file is imported by default into the program. A new file is to be generated whenever analyzing a new genome.

### MultiFASTA protein sequence file












